tmt/steps/provision/libvirt.py

```diff
@@ -0,0 +1,157 @@
+import re
+from pathlib import Path
+from time import sleep
+from typing import Tuple
+
+import click
+
+import tmt
+
+VM_CLONE_OUT_RE = re.compile(".+'(.+)' created successfully")
+VM_IP_RE = re.compile(".+ipv4\\s+(.+)\\/\\d+")
+VM_WAITING_COUNTER_MAX_NET = 30
+
+
+class ProvisionLibvirt(tmt.steps.provision.ProvisionPlugin):
+    """Libvirt provision plugin.
+
+    Prerequisite to have the vm already prepared via libvirt.
+    Plugin clones the existing VM and do the rest steps.
+
+    Example config:
+    ```yaml
+    provision:
+        how: libvirt
+        origin_vm_name: c2r_centos7_template
+        develop: true
+    ```
+
+    Needed packages:
+      virt-install \
+      virt-manager \
+      virt-viewer \
+      virt-viewer \
+      qemu-kvm \
+      libvirt-devel \
+      libvirt-daemon-config-network \
+      libvirt-daemon-kvm \
+      libguestfs-tools
+    """
+
+    # Guest instance
+    _guest = None
+
+    # Supported keys
+    _keys = ["origin_vm_name", "develop"]
```

Comment:
`_keys` are no longer the thing, same applies to the use of `data` in guest instance creation, it's no longer a dictionary. Instead, we're using Python's data classes, see https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L172 for guest "data" packages, and https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L187 for plugin's own (it's usually the same, hence the thin wrapping).

Then there's a pointer from the guest to its data class, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L205, plugin referencing its data class, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L559, and plugin referencing its guest class, https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L560.